### PR TITLE
Optimize CPU RNN performance with MKL-DNN

### PR DESCRIPTION
**Pull request by JianpingChen066**
This is the code rebase for PR15914
More detailed description, please have a look at 
https://github.com/pytorch/pytorch/pull/15914

### CODE REVIEW

**Comment by gchanan** [comment]
do these work with non-contiguous tensors?  See https://github.com/pytorch/pytorch/pull/16301 for the issue with mkldnn convolution on non-contiguous tensors.

**Comment by cpuhrsch** [comment]
There's been [recent changes](https://github.com/pytorch/pytorch/pull/16239) to the function schema used in native_functions.yaml. Please rebase again. You can use the following patch to resolve the build errors.

```
diff -r 69c786a8ff0c -r 1c24f8b0c777 aten/src/ATen/native/native_functions.yaml
--- a/aten/src/ATen/native/native_functions.yaml  Wed Jan 30 10:47:29 2019 -0800
+++ b/aten/src/ATen/native/native_functions.yaml  Wed Jan 30 12:59:21 2019 -0800
@@ -1423,17 +1423,17 @@
 - func: mkldnn_convolution_backward(Tensor self, Tensor grad_output, Tensor weight, int[] padding, int[] stride, int[] dilation, int groups, bool[3] output_mask) -> (Tensor, Tensor, Tensor)
   matches_jit_signature: True

-- func: rnn_tanh_mkldnn_stub(Tensor self, Tensor hx, TensorList params, bool has_biases, int64_t num_layers, double dropout, bool train, bool bidirectional, bool batch_first) -> (Tensor, Tensor)
+- func: rnn_tanh_mkldnn_stub(Tensor self, Tensor hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, bool batch_first) -> (Tensor, Tensor)

-- func: rnn_relu_mkldnn_stub(Tensor self, Tensor hx, TensorList params, bool has_biases, int64_t num_layers, double dropout, bool train, bool bidirectional, bool batch_first) -> (Tensor, Tensor)
+- func: rnn_relu_mkldnn_stub(Tensor self, Tensor hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, bool batch_first) -> (Tensor, Tensor)

-- func: gru_mkldnn_stub(Tensor self, Tensor hx, TensorList params, bool has_biases, int64_t num_layers, double dropout, bool train, bool bidirectional, bool batch_first) -> (Tensor, Tensor)
+- func: gru_mkldnn_stub(Tensor self, Tensor hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, bool batch_first) -> (Tensor, Tensor)

-- func: lstm_mkldnn_stub(Tensor self, TensorList hx, TensorList params, bool has_biases, int64_t num_layers, double dropout, bool train, bool bidirectional, bool batch_first) -> (Tensor, Tensor, Tensor)
+- func: lstm_mkldnn_stub(Tensor self, Tensor[] hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, bool batch_first) -> (Tensor, Tensor, Tensor)

-- func: mkldnn_rnn(Tensor input, TensorList weight, int64_t weight_stride0, Tensor hx, Tensor? cx, int64_t mode, int64_t hidden_size, int64_t num_layers, bool batch_first, double dropout, bool train, bool bidirectional, IntList batch_sizes, BoolTensor? dropout_state) -> (Tensor, Tensor, Tensor, TensorList)
+- func: mkldnn_rnn(Tensor input, Tensor[] weight, int weight_stride0, Tensor hx, Tensor? cx, int mode, int hidden_size, int num_layers, bool batch_first, float dropout, bool train, bool bidirectional, int[] batch_sizes, BoolTensor? dropout_state) -> (Tensor, Tensor, Tensor, Tensor[])

-- func: mkldnn_rnn_backward(Tensor input, TensorList weight, int64_t weight_stride0, Tensor hx, Tensor? cx, Tensor output, Tensor? grad_output, Tensor? grad_hy, Tensor? grad_cy, int64_t mode, int64_t hidden_size, int64_t num_layers, bool batch_first, double dropout, bool train, bool bidirectional, IntList batch_sizes, BoolTensor? dropout_state, TensorList reserve, std::array<bool,4> output_mask) -> (Tensor, Tensor, Tensor, TensorList)
+- func: mkldnn_rnn_backward(Tensor input, Tensor[] weight, int weight_stride0, Tensor hx, Tensor? cx, Tensor output, Tensor? grad_output, Tensor? grad_hy, Tensor? grad_cy, int mode, int hidden_size, int num_layers, bool batch_first, float dropout, bool train, bool bidirectional, int[] batch_sizes, BoolTensor? dropout_state, Tensor[] reserve, bool[4] output_mask) -> (Tensor, Tensor, Tensor, Tensor[])

 - func: miopen_batch_norm(Tensor input, Tensor weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, float exponential_average_factor, float epsilon) -> (Tensor, Tensor, Tensor)
   matches_jit_signature: True
```

**Comment by JianpingChen066** [comment]
@gchanan Thank you for raising this issue, we will have a further investigate and enhancement on this

File: aten/src/ATen/native/RNN.cpp [hunk 1]
```diff
@@ -665,6 +670,15 @@ std::tuple<Tensor, Tensor, Tensor> lstm(
             num_layers, dropout_p, train, bidirectional, batch_first);
     return std::make_tuple(output, hy, cy);
   }
+  if (at::mkldnn_is_acceptable(_input)
+      && (!train || (train && _input.size(2) == hx[0].size(2) && !bidirectional))) {
+    // Disable MKLDNN RNN for training when input_size != hidden_size and bidirectional
+    // since workspace offset is miss calculated on such cases
+    // This issue has been fixed on https://github.com/intel/mkl-dnn since commit d5c4eb5e
```
Comment (apaszke) [inline aten/src/ATen/native/RNN.cpp]: If it's been fixed, can't we simply use that commit?

Comment (JianpingChen066) [reply aten/src/ATen/native/RNN.cpp]: @apaszke 
Current mkl-dnn version under pytorch/third_party/ideep/mkl-dnn still not upgraded after the commit version d5c4eb5e, so the issue is still exist and you can simply use this commit as the workaround for this. Thanks!

Comment (apaszke) [reply aten/src/ATen/native/RNN.cpp]: Sure. My question was if it was possible to update it?

**Comment by JianpingChen066** [comment]
@apaszke
for MKLDNN update, it may need more validation on this to avoid  unnecessary errors or regression to Pytorch users. So may need sometime for it. Thanks

**Comment by JianpingChen066** [comment]
@mingfeima @Jianhui-Li
I will close this to avoid duplication, Mingfei, you can have a further rebase on #15914. Thanks

**Pull request closed without merging**